src/main.jl

```diff
@@ -391,16 +394,23 @@ function resolve(goals::Vector{<:Term}, clauses::ClauseTable{T}; options...)  wh
         # Iterate across clause set with matching heads
         matched_clauses = retrieve_clauses(clauses, term, funcs)
         matched = false
+        local_queue = MutableLinkedList{GoalTree}() # maintain a local queue for the depth first search; the new goals should be added to the global queue in this order to the front
         for c in matched_clauses
             # Freshen variables in clause
             c = freshen!(c, Subst(), vcount)
             # If term unifies with head of a clause, add it as a subgoal
             unifier = unify(term, c.head, occurs_check, funcs)
             if isnothing(unifier) continue end
             child = GoalTree(c.head, goal, copy(c.body), 1, unifier, vmap)
-            push!(queue, child)
+            (search==:dfs) ? push!(local_queue, child) : push!(queue, child)
             matched = true
         end
+
+        #update the global queue with the local one
+        while (search==:dfs) && length(local_queue) > 0
+            pushfirst!(queue, pop!(local_queue))
+        end
+        
```

Comment:
Instead of pushing each child to the local queue, then transferring the items to the global queue, why we just `pushfirst!` directly to the global queue?

Reply:
Oh, after giving it more thought, I think I understand why doing it this way leads to different behavior.

Rather than using a local queue though, I think what we can do is replace the line:
```julia
for c in matched_clauses
```

with 

```julia
for c in Iterators.reverse(matched_clauses)
```

This will iterate over the matched clauses in reverse order, allowing us to `pushfirst!` directly to the global queue.

Reply:
you are right, this is a much better solution! Better not to change data structures if not necessary.
I was searching for the most efficient data structure for front/back adding, but I like this solution much more.